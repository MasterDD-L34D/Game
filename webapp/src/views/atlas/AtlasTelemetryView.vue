<template>
  <section class="atlas-telemetry" aria-live="polite">
    <header class="atlas-telemetry__header">
      <div>
        <h2>Telemetria Nebula</h2>
        <p class="atlas-telemetry__status" :data-mode="telemetryStatus.mode">
          {{ telemetryStatus.label }} · {{ datasetStatus.label }}
        </p>
      </div>
      <div class="atlas-telemetry__actions">
        <button type="button" @click="refresh">Aggiorna ora</button>
        <button type="button" @click="activateDemo">Modalità demo</button>
      </div>
    </header>

    <p v-if="moduleError" class="atlas-telemetry__error">{{ moduleError }}</p>

    <section v-else class="atlas-telemetry__grid">
      <article class="atlas-telemetry__panel">
        <h3>Riepilogo eventi</h3>
        <ul class="atlas-telemetry__summary">
          <li>
            <span>Eventi totali</span>
            <strong>{{ telemetrySummary.total }}</strong>
          </li>
          <li>
            <span>Eventi aperti</span>
            <strong :data-tone="telemetrySummary.open > 0 ? 'warning' : 'success'">
              {{ telemetrySummary.open }}
            </strong>
          </li>
          <li>
            <span>Priorità alta</span>
            <strong :data-tone="telemetrySummary.highPriority > 0 ? 'critical' : 'neutral'">
              {{ telemetrySummary.highPriority }}
            </strong>
          </li>
          <li>
            <span>Acknowledged</span>
            <strong>{{ telemetrySummary.acknowledged }}</strong>
          </li>
        </ul>
        <p class="atlas-telemetry__updated">Ultimo evento: {{ telemetrySummary.lastEventLabel }}</p>
      </article>

      <article class="atlas-telemetry__panel">
        <h3>Distribuzione readiness</h3>
        <ul class="atlas-telemetry__chips">
          <li v-for="chip in readinessChips" :key="chip.id" :data-tone="chip.tone">
            <span>{{ chip.label }}</span>
            <strong>{{ chip.value }}</strong>
            <small v-if="chip.badge">{{ chip.badge }}</small>
          </li>
        </ul>
      </article>

      <article class="atlas-telemetry__panel atlas-telemetry__panel--chart">
        <header>
          <h3>Copertura QA</h3>
          <span>{{ telemetryCoverageAverage }}%</span>
        </header>
        <SparklineChart :points="coverageStream" color="#4f46e5" />
      </article>

      <article class="atlas-telemetry__panel atlas-telemetry__panel--chart">
        <header>
          <h3>Timeline incidenti</h3>
          <span>7 giorni</span>
        </header>
        <SparklineChart :points="incidentStream" color="#0ea5e9" />
      </article>

      <article class="atlas-telemetry__panel atlas-telemetry__panel--chart">
        <header>
          <h3>High priority</h3>
          <span>Ticket giornalieri</span>
        </header>
        <SparklineChart :points="highPriorityStream" color="#f97316" />
      </article>
    </section>
  </section>
</template>

<script setup>
import { computed } from 'vue';

import SparklineChart from '../../components/metrics/SparklineChart.vue';
import { useNebulaProgressModule } from '../../modules/useNebulaProgressModule';

const moduleState = useNebulaProgressModule();

const datasetStatus = computed(() => {
  const status = moduleState.datasetStatus?.value;
  if (status) {
    return status;
  }
  return { label: 'Dataset live', source: 'remote', demo: false };
});

const telemetryStatus = computed(() => moduleState.telemetryStatus.value);
const moduleError = computed(() => moduleState.error.value?.message || null);

const telemetrySummary = computed(() => moduleState.telemetrySummary.value);
const telemetryCoverageAverage = computed(() => moduleState.telemetryCoverageAverage.value);
const coverageStream = computed(() => moduleState.telemetryStreams.value.coverage);
const incidentStream = computed(() => moduleState.telemetryStreams.value.incidents);
const highPriorityStream = computed(() => moduleState.telemetryStreams.value.highPriority);

const readinessChips = computed(() => {
  const distribution = moduleState.telemetryDistribution.value;
  const status = telemetryStatus.value;
  const dataset = datasetStatus.value;
  const chips = [
    { id: 'success', label: 'Pronte', tone: 'success', value: distribution.success },
    { id: 'warning', label: 'In attesa', tone: 'warning', value: distribution.warning },
    { id: 'neutral', label: 'Neutrali', tone: 'neutral', value: distribution.neutral },
    { id: 'critical', label: 'Bloccate', tone: 'critical', value: distribution.critical },
  ];
  if (status.offline || dataset.source !== 'remote' || dataset.demo) {
    const badge = status.mode === 'mock' ? 'Telemetria demo' : 'Fallback dataset';
    return chips.map((chip) => ({ ...chip, badge }));
  }
  return chips.map((chip) => ({ ...chip, badge: '' }));
});

const refresh = () => moduleState.refresh();
const activateDemo = () => {
  if (typeof moduleState.activateDemoTelemetry === 'function') {
    return moduleState.activateDemoTelemetry();
  }
  return moduleState.refresh();
};
</script>

<style scoped>
.atlas-telemetry {
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.atlas-telemetry__header {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  gap: 1rem;
}

.atlas-telemetry__status {
  font-size: 0.95rem;
  color: #334155;
}

.atlas-telemetry__status[data-mode='fallback'] {
  color: #b45309;
}

.atlas-telemetry__status[data-mode='mock'] {
  color: #6366f1;
}

.atlas-telemetry__actions {
  display: flex;
  gap: 0.75rem;
}

.atlas-telemetry__actions button {
  padding: 0.6rem 1rem;
  border-radius: 0.75rem;
  border: 1px solid transparent;
  background: #1e293b;
  color: #fff;
  font-weight: 600;
  cursor: pointer;
}

.atlas-telemetry__actions button:last-child {
  background: transparent;
  color: #1e293b;
  border-color: rgba(30, 41, 59, 0.2);
}

.atlas-telemetry__error {
  padding: 1rem 1.5rem;
  border-radius: 1rem;
  background: rgba(248, 113, 113, 0.12);
  color: #991b1b;
}

.atlas-telemetry__grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(16rem, 1fr));
  gap: 1.5rem;
}

.atlas-telemetry__panel {
  padding: 1.5rem;
  border-radius: 1.25rem;
  background: #fff;
  box-shadow: 0 12px 32px rgba(15, 23, 42, 0.08);
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.atlas-telemetry__summary {
  list-style: none;
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(8rem, 1fr));
  gap: 0.75rem;
}

.atlas-telemetry__summary li {
  display: flex;
  flex-direction: column;
  gap: 0.35rem;
  color: #475569;
}

.atlas-telemetry__summary strong {
  font-size: 1.4rem;
  color: #0f172a;
}

.atlas-telemetry__summary strong[data-tone='warning'] {
  color: #b45309;
}

.atlas-telemetry__summary strong[data-tone='critical'] {
  color: #b91c1c;
}

.atlas-telemetry__updated {
  font-size: 0.9rem;
  color: #475569;
}

.atlas-telemetry__chips {
  list-style: none;
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(10rem, 1fr));
  gap: 0.75rem;
}

.atlas-telemetry__chips li {
  padding: 0.85rem 1rem;
  border-radius: 1rem;
  background: rgba(15, 23, 42, 0.04);
  display: grid;
  grid-template-columns: 1fr auto;
  gap: 0.25rem;
  align-items: center;
}

.atlas-telemetry__chips li[data-tone='success'] {
  background: rgba(34, 197, 94, 0.14);
}

.atlas-telemetry__chips li[data-tone='warning'] {
  background: rgba(234, 179, 8, 0.16);
}

.atlas-telemetry__chips li[data-tone='critical'] {
  background: rgba(248, 113, 113, 0.18);
}

.atlas-telemetry__chips li strong {
  justify-self: end;
  font-size: 1.2rem;
  color: #0f172a;
}

.atlas-telemetry__chips li small {
  grid-column: 1 / -1;
  font-size: 0.75rem;
  color: #475569;
}

.atlas-telemetry__panel--chart header {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
  font-size: 0.95rem;
  color: #334155;
}
</style>
