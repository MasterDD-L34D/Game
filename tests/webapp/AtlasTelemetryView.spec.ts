import { describe, expect, it, beforeEach, vi } from 'vitest';
import { computed, ref } from 'vue';
import { mount } from '@vue/test-utils';

import AtlasTelemetryView from '../../webapp/src/views/atlas/AtlasTelemetryView.vue';

const refreshMock = vi.fn();
const activateDemoMock = vi.fn();

vi.mock('../../webapp/src/modules/useNebulaProgressModule', () => ({
  useNebulaProgressModule: () => ({
    datasetStatus: computed(() => ({
      source: 'remote',
      label: 'Dataset live',
      demo: false,
    })),
    telemetryStatus: computed(() => ({
      mode: 'live',
      label: 'Telemetria live',
      offline: false,
      demo: false,
    })),
    error: computed(() => ({ message: null })),
    telemetrySummary: computed(() => ({
      total: 12,
      open: 3,
      acknowledged: 7,
      highPriority: 2,
      lastEventLabel: 'Aggiornato 5m fa',
    })),
    telemetryStreams: computed(() => ({
      coverage: [62, 68, 74, 81],
      incidents: [1, 0, 2, 1],
      highPriority: [0, 1, 1, 0],
    })),
    telemetryDistribution: computed(() => ({
      success: 5,
      warning: 2,
      neutral: 3,
      critical: 2,
    })),
    telemetryCoverageAverage: computed(() => 78),
    generatorStatus: computed(() => ({
      status: 'success',
      label: 'Generatore online',
      generatedAt: '2035-04-01T12:00:00Z',
    })),
    generatorMetrics: computed(() => ({
      generationTimeMs: 512,
      speciesTotal: 24,
      enrichedSpecies: 16,
      eventTotal: 4,
      datasetSpeciesTotal: 40,
      coverageAverage: 78,
      coreTraits: 48,
      optionalTraits: 22,
      synergyTraits: 12,
    })),
    generatorStreams: computed(() => ({
      generationTime: [420, 480, 512],
      species: [18, 22, 24],
      enriched: [10, 14, 16],
    })),
    refresh: refreshMock,
    activateDemoTelemetry: activateDemoMock,
    loading: ref(false),
    lastUpdated: ref('2035-04-01T12:00:00Z'),
  }),
}));

describe('AtlasTelemetryView', () => {
  beforeEach(() => {
    refreshMock.mockClear();
    activateDemoMock.mockClear();
  });

  it('visualizza metriche principali e consente il refresh manuale', async () => {
    const wrapper = mount(AtlasTelemetryView, {
      global: {
        stubs: {
          SparklineChart: {
            props: ['points'],
            template: "<div class=\"sparkline-stub\">{{ points.join(',') }}</div>",
          },
        },
      },
    });

    expect(wrapper.find('h2').text()).toBe('Telemetria Nebula');
    expect(wrapper.text()).toContain('Eventi totali');
    expect(wrapper.text()).toContain('Distribuzione readiness');

    const [refreshButton, demoButton] = wrapper.findAll('button');
    await refreshButton.trigger('click');
    await demoButton.trigger('click');

    expect(refreshMock).toHaveBeenCalledTimes(1);
    expect(activateDemoMock).toHaveBeenCalledTimes(1);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
