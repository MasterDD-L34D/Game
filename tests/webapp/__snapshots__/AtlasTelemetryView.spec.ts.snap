// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AtlasTelemetryView > visualizza metriche principali e consente il refresh manuale 1`] = `
"<section data-v-386358bc="" class="atlas-telemetry" aria-live="polite">
  <header data-v-386358bc="" class="atlas-telemetry__header">
    <div data-v-386358bc="">
      <h2 data-v-386358bc="">Telemetria Nebula</h2>
      <p data-v-386358bc="" class="atlas-telemetry__status" data-mode="live">Telemetria live · Dataset live</p>
    </div>
    <div data-v-386358bc="" class="atlas-telemetry__actions"><button data-v-386358bc="" type="button">Aggiorna ora</button><button data-v-386358bc="" type="button">Modalità demo</button></div>
  </header>
  <section data-v-386358bc="" class="atlas-telemetry__grid">
    <article data-v-386358bc="" class="atlas-telemetry__panel">
      <h3 data-v-386358bc="">Riepilogo eventi</h3>
      <ul data-v-386358bc="" class="atlas-telemetry__summary">
        <li data-v-386358bc=""><span data-v-386358bc="">Eventi totali</span><strong data-v-386358bc="">12</strong></li>
        <li data-v-386358bc=""><span data-v-386358bc="">Eventi aperti</span><strong data-v-386358bc="" data-tone="warning">3</strong></li>
        <li data-v-386358bc=""><span data-v-386358bc="">Priorità alta</span><strong data-v-386358bc="" data-tone="critical">2</strong></li>
        <li data-v-386358bc=""><span data-v-386358bc="">Acknowledged</span><strong data-v-386358bc="">7</strong></li>
      </ul>
      <p data-v-386358bc="" class="atlas-telemetry__updated">Ultimo evento: Aggiornato 5m fa</p>
    </article>
    <article data-v-386358bc="" class="atlas-telemetry__panel">
      <h3 data-v-386358bc="">Distribuzione readiness</h3>
      <ul data-v-386358bc="" class="atlas-telemetry__chips">
        <li data-v-386358bc="" data-tone="success"><span data-v-386358bc="">Pronte</span><strong data-v-386358bc="">5</strong>
          <!--v-if-->
        </li>
        <li data-v-386358bc="" data-tone="warning"><span data-v-386358bc="">In attesa</span><strong data-v-386358bc="">2</strong>
          <!--v-if-->
        </li>
        <li data-v-386358bc="" data-tone="neutral"><span data-v-386358bc="">Neutrali</span><strong data-v-386358bc="">3</strong>
          <!--v-if-->
        </li>
        <li data-v-386358bc="" data-tone="critical"><span data-v-386358bc="">Bloccate</span><strong data-v-386358bc="">2</strong>
          <!--v-if-->
        </li>
      </ul>
    </article>
    <article data-v-386358bc="" class="atlas-telemetry__panel atlas-telemetry__panel--chart">
      <header data-v-386358bc="">
        <h3 data-v-386358bc="">Copertura QA</h3><span data-v-386358bc="">78%</span>
      </header>
      <div data-v-386358bc="" class="sparkline-stub" color="#4f46e5">62,68,74,81</div>
    </article>
    <article data-v-386358bc="" class="atlas-telemetry__panel atlas-telemetry__panel--chart">
      <header data-v-386358bc="">
        <h3 data-v-386358bc="">Timeline incidenti</h3><span data-v-386358bc="">7 giorni</span>
      </header>
      <div data-v-386358bc="" class="sparkline-stub" color="#0ea5e9">1,0,2,1</div>
    </article>
    <article data-v-386358bc="" class="atlas-telemetry__panel atlas-telemetry__panel--chart">
      <header data-v-386358bc="">
        <h3 data-v-386358bc="">High priority</h3><span data-v-386358bc="">Ticket giornalieri</span>
      </header>
      <div data-v-386358bc="" class="sparkline-stub" color="#f97316">0,1,1,0</div>
    </article>
  </section>
</section>"
`;
